src/D2L.CodeStyle.Analyzers/CustomerState/CustomerStateAnalyzer.cs

```diff
@@ -0,0 +1,262 @@
+﻿using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.CustomerState {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class CustomerStateAnalyzer : DiagnosticAnalyzer {
+
+		public enum CustomerStateResult {
+			Unknown,
+
+			NoCustomerState,
+
+			CustomerState
+		}
+
+		public class CustomerStateAnalysis {
+			public CustomerStateResult State { get; set; }
+
+			public bool IsHidden { get; set; }
+		}
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.SingletonDependencyHasCustomerState,
+			Diagnostics.PublicClassHasHiddenCustomerState
+		);
+
+		public override void Initialize(
+			AnalysisContext context
+		) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis(
+			CompilationStartAnalysisContext context
+		) {
+			var singletonAttribute = context.Compilation.GetTypeByMetadataName(
+				"D2L.LP.Extensibility.Activation.Domain.SingletonAttribute" );
+
+			var customerStateAttribute = context.Compilation.GetTypeByMetadataName(
+				"D2L.LP.Extensibility.Activation.Domain.CustomerStateAttribute" );
+
+			if( singletonAttribute == default( INamedTypeSymbol )

```

Comment:
There's an extension method `.IsNullOrErrorType()` if that's the behaviour you're looking for.